crates/ty_python_semantic/src/types.rs

```diff
@@ -6911,6 +6952,94 @@ impl<'db> BoundMethodType<'db> {
         ))
     }
 
+    /// Infers this method scope's types and returns the inferred return type.
+    #[salsa::tracked(cycle_fn=method_return_type_cycle_recover, cycle_initial=method_return_type_cycle_initial)]
+    pub(crate) fn inferred_return_type(self, db: &'db dyn Db) -> Type<'db> {
+        let scope = self
+            .function(db)
+            .literal(db)
+            .last_definition(db)
+            .body_scope(db);
+        let inference = infer_scope_types(db, scope);
+        inference.inferred_return_type(db, Some(self))
+    }
+
+    pub(crate) fn is_final(self, db: &'db dyn Db) -> bool {
+        if self
+            .function(db)
+            .has_known_decorator(db, FunctionDecorators::FINAL)
+        {
+            return true;
+        }
+        let definition_scope = self.function(db).definition(db).scope(db);
+        let index = semantic_index(db, definition_scope.file(db));
+        let module = parsed_module(db.upcast(), definition_scope.file(db)).load(db.upcast());
+        let class_definition =
+            index.expect_single_definition(definition_scope.node(db).expect_class(&module));
+        let class_ty = binding_type(db, class_definition).expect_class_literal();
+        class_ty
+            .known_function_decorators(db)
+            .any(|deco| deco == KnownFunction::Final)
+    }
+
+    /// Returns the compatible return type for this method -- the intersection of the return types of the base class methods.
+    pub(crate) fn compatible_return_type(self, db: &'db dyn Db) -> Option<Type<'db>> {
```

Comment:
Given the above comments on the tests, I think we don't need this method -- it can be replaced by a simple linear MRO lookup.